source/cppfront.cpp

```diff
@@ -1928,6 +1933,18 @@ class cppfront
         }
         suppress_move_from_last_use = false;
 
+        if (args) {
+            // if after printing core expression args is defined
+            // it means that the chaining started by function call
+            // we need to print its arguments
+            suffix.emplace_back(")", n.position());
+            for (auto&& e: args.value()) {
+                suffix.push_back(e);
+            }
+            suffix.emplace_back("(", n.position());
```

Comment:
I think these two lines that print `"("` and `")"` may be what's causing the whitespace changes I noticed in some of the regression tests. Should we extend `args` to store the positions of the left and right parens, and then use those positions here? Just using `n.position()` is using the position of the whole expression and we're past that... the parens may even be on other lines, as they are in `mixed-function-expression-and-std-for-each.cpp2` and `mixed-postfix-expression-custom-formatting.cpp2`... those are two examples where I saw the `)` get moved up/left instead of keeping its original position.

Reply:
Yes, you are right. I struggled to find a solution as I know it is essential for you.

I tried to use `push_need_expression_list_parens(true)` at the generation of args, but something was wrong. I will test it once again.

Reply:
You shouldn't have to push the parens state... The basic idea would be to store the open and close paren positions alongside `args` itself, either as `source_position`s or as `token const*`s to the paren tokens so we can ask them their positions.

But I can also go do this post-merge if that's the only thing holding up the merge, don't let this alone block the PR.

Reply:
OK, I've found and tested the solution to correct the paren positions. I'll post it here as another review, it's just two more lines to remember the `(` and `)` position and then using them in the three places.